python/ray/data/_internal/datasource/clickhouse_datasource.py

```diff
@@ -0,0 +1,201 @@
+import logging
+from typing import Any, Dict, List, Optional, Tuple
+
+from ray.data._internal.util import _check_import
+from ray.data.block import Block, BlockMetadata
+from ray.data.datasource.datasource import Datasource, ReadTask
+from ray.util.annotations import DeveloperAPI
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class ClickHouseDatasource(Datasource):
+    """
+    A Ray datasource for reading from ClickHouse.
+    """
+
+    def __init__(
+        self,
+        entity: str,
+        dsn: str,
+        columns: Optional[List[str]] = None,
+        filters: Optional[Dict[str, Tuple[str, Any]]] = None,
+        order_by: Optional[Tuple[List[str], bool]] = None,
+        client_settings: Optional[Dict[str, Any]] = None,
+        client_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Initialize a ClickHouseDatasource.
+
+        Args:
+            entity: Fully qualified table or view identifier (e.g.,
+                "default.table_name").
+            dsn: A string in standard DSN (Data Source Name) format.
+            columns: Optional list of columns to select from the data source.
+                If no columns are specified, all columns will be selected by default.
+            filters: Optional fields and values mapping to use to filter the data via
+                WHERE clause. The value should be a tuple where the first element is
+                one of ('is', 'not', 'less', 'greater') and the second
+                element is the value to filter by. The default operator
+                is 'is'. Only strings, ints, floats, booleans,
+                and None are allowed as values.
```

Comment:
IIUC this is requiring predicate in DNF format, let's call it out explicitly and add an example to help with understanding of it. 

Also let's add a link to the page of parameters to ClickHouse explaining these in more details

Reply:
I attempted to resolve this in my latest commit. I added an example, included a link to ClickHouse documentation, and went much deeper into details.

One item I wanted to callout was `filters` currently only supports joining by `AND` operators. My thinking was as follows:

1. I'm assuming for the vast majority of use-cases that feature engineering work will be done in ClickHouse and the end user would simply want to bring in data from a view. I didn't see the need to build an extensive query builder without it being necessary.
2. The main purpose of the filters was to offer the end user a way to reduce the data being transferred into Ray via a simple filtering mechanism.

I did leave myself a TODO to add support for filtering by datetime types in a future PR. Could I also add support for `OR` operators along with defining a DNF format in a future PR if necessary?

Reply:
@jecsand838 yes, totally we can make this a follow-up. 

There are a few things i want to call out though:

 - We're currently adding support for generic expressions (to enable future advanced optimizations powered by it) and therefore want to make sure that we consolidate all expression handling onto a single engine (@richardliaw is working on a PR as we speak)
 - In the meantime, we also need to make sure we're not flip-flopping on APIs back and forth between releases

As such, i'd recommended we extract adding filtering push-down into a separate PR (stacked on top of this one) that we can do one more iteration on to consolidate the expression handling before we put it out for everyone to use.

Does that make sense?

Reply:
@alexeykudinkin Makes complete sense, I'll take care of that.

Reply:
@alexeykudinkin The filtering functionality has been extracted from this PR and placed here: https://github.com/jecsand838/ray/pull/1